cassandra/cluster.py

```diff
@@ -721,18 +741,17 @@ def shutdown(self):
             else:
                 self.is_shutdown = True
 
-        if self.scheduler:
-            self.scheduler.shutdown()
+        self.scheduler.shutdown()
 
-        if self.control_connection:
-            self.control_connection.shutdown()
+        self.control_connection.shutdown()
 
-        if self.sessions:
-            for session in self.sessions:
-                session.shutdown()
+        for session in self.sessions:
+            session.shutdown()
+
+        self.executor.shutdown()
 
-        if self.executor:
-            self.executor.shutdown()
+        if self._idle_heartbeat:
+            self._idle_heartbeat.stop()
```

Comment:
It shouldn't make a big difference, but I would stop this before closing the control conn and sessions.